### PR TITLE
Add phantom readers based on the given set of policies

### PR DESCRIPTION
**Pull request by bgogul**
Instead of creating actual phantom reader recipes, this PR _simulates_ phantom readers during ingress restricting.

### CODE REVIEW

File: src/tools/tests/allocator-recipe-resolver-test.ts [hunk 1]
```diff
@@ -1229,7 +1398,7 @@ policy Policy0 {
   }
 }
 policy Policy1 {
-  @allowedRetention(medium: 'Ram', encryption: true)
+  @allowedRetention(medium: 'Ram', encryption: false)
```
Comment (mmandlis) [inline src/tools/tests/allocator-recipe-resolver-test.ts:1401]: could you, please, explain why these capabilities needed to change?

File: src/tools/tests/allocator-recipe-resolver-test.ts [hunk 2]
```diff
@@ -1149,6 +1157,167 @@ recipe ReadingRecipeAD
       ${recipe}
     `, 'Thing {a: (Text, Number), c: List<(Text, Number)>, b: [(Text, Number)], d: (Text, Number)}');
   });
+
+  it.only('restricts writer fields with phantom readers', async () => {
+    const verifyPhantomReader = async (
```
Comment (mmandlis) [inline src/tools/tests/allocator-recipe-resolver-test.ts:1162]: suggestion - maybe factor out the `verifyPhantomReader` method outside the test case and break the test case to smaller more focused ones?

File: src/tools/tests/allocator-recipe-resolver-test.ts [hunk 3]
```diff
@@ -1149,6 +1157,167 @@ recipe ReadingRecipeAD
...
+      async () => verifyPhantomReader(
+        thingSchemaWithTuples,
+        'from Thing access { a }',
+        'Thing',
+        'Thing {a: (Text, Number)}'
+      ),
+      FieldPathError,
+      `Unsupported field type`
+    );
+  });
```
Comment (mmandlis) [inline src/tools/tests/allocator-recipe-resolver-test.ts:1319]: could you add a test case for the error case - `Unable to find max read type for handle`?
maybe also a case when the writer fields' and the policy fields intersection is empty?

**Review by mmandlis — APPROVED** [review]

File: src/tools/tests/allocator-recipe-resolver-test.ts [hunk 4]
```diff
@@ -1149,6 +1157,167 @@ recipe ReadingRecipeAD
...
+        a: Text
+        b: Text
+        c: Text
+        d: Text
+        e: Text
+    `;
+    await verifyPhantomReader(
+      thingSchema,
+      'from Thing access {a, b, d}',
+      'Thing',
```
Comment (piotrswigon) [inline src/tools/tests/allocator-recipe-resolver-test.ts:1200]: The best practice is to use inline schemas in particles. The benefit here would also be that you show what's written, and what's expected next to each other.

E.g.

```
    await verifyPhantomReader(
      thingSchema,
      'from Thing access {a, b, d}',
      'Thing {a, b, c, d, e}',
      'Thing {a: Text, b: Text, d: Text}'
    );
```

I would maybe refactor the method for readability like this:
```
    await verifyPhantomReader({
      schema: thingSchema,
      policy: 'from Thing access {a, b, d}',
      written: 'Thing {a, b, c, d, e}',
      expected: 'Thing {a: Text, b: Text, d: Text}'
    });
```

I've seen that Josh has been working to enable { } at the schema level, so maybe you could even inline reasonably:

```
    await verifyPhantomReader({
      schema: 'schema Thing {a: Text, b: Text, c:Text}',
      policy: 'from Thing access {a, c}',
      written: 'Thing {a, b, c}',
      expected: 'Thing {a: Text, c: Text}'
    });
```

File: src/tools/tests/allocator-recipe-resolver-test.ts [hunk 5]
```diff
@@ -1149,6 +1157,167 @@ recipe ReadingRecipeAD
...
+      `from Thing access {
+         foo {a, d},
+         bar
+      }`,
+      'Thing',
+      'Thing {foo: List<inline Foo {a: Text, d: Text}>, bar: Text}'
+    );
+    const thingSchemaWithTuples = `
+      schema Thing
+        a: (Text, Number)
```
Comment (piotrswigon) [inline src/tools/tests/allocator-recipe-resolver-test.ts:1295]: Ah, tuples, when will they be supported E2E? You're forward looking 🥇

**Review by piotrswigon — APPROVED** [review]
+1 to Maria's: why capabilities had to change?

File: src/tools/allocator-recipe-resolver.ts [hunk 6]
```diff
@@ -122,6 +122,25 @@ export class AllocatorRecipeResolver {
     for (const handleId of Object.keys(handleById)) {
       const {store, handles} = handleById[handleId];
       const allTypes = handles.map(h => h.type);
+      if (this.ingressValidation) {
```
[the remark this answers is not in the document]
Comment (bgogul) [reply src/tools/allocator-recipe-resolver.ts:125]: Filed a bug and will take care of it shortly.

File: src/tools/tests/allocator-recipe-resolver-test.ts (hunk 5 above)
Comment (bgogul) [reply src/tools/tests/allocator-recipe-resolver-test.ts:1364]: :)

File: src/tools/tests/allocator-recipe-resolver-test.ts (hunk 1 above)
Comment (bgogul) [reply src/tools/tests/allocator-recipe-resolver-test.ts:1491]: With encryption, there was no storage driver and allocator was not able to create a storage key.

File: src/tools/tests/allocator-recipe-resolver-test.ts (hunk 3 above)
Comment (bgogul) [reply src/tools/tests/allocator-recipe-resolver-test.ts:1388]: Done. I added the following tests:
 - `Unable to find max read type` error.
 - example where intersection of writer fields and policy fields is empty.

**Review by bgogul — COMMENTED** [review]
Took care of the comments and tests.

**Comment by bgogul** [comment]
> +1 to Maria's: why capabilities had to change?

There is no storage driver with encryption support and therefore, allocator was not able to create storage key.

File: src/tools/tests/allocator-recipe-resolver-test.ts (hunk 4 above)
Comment (bgogul) [reply src/tools/tests/allocator-recipe-resolver-test.ts]: Thanks for the suggestion, Piotr. I have refactored the tests for readability.

File: src/tools/tests/allocator-recipe-resolver-test.ts (hunk 2 above)
Comment (bgogul) [reply src/tools/tests/allocator-recipe-resolver-test.ts]: Thanks, Maria! I have split the tests into smaller focussed ones.

**Pull request closed without merging**